api/envoy/extensions/filters/http/grpc_field_extraction/v3/config.proto

```diff
@@ -0,0 +1,137 @@
+// type and its value will be extracted in string format.
+//
+//
+// The output is a serialized google.protobuf.Struct of field<StringValue> -> extract values<ListValue of StringValue>. It will be injected in the `header` with base64 encoded format.
+//
+// For example, we have the following request definition for the gRPC method `/pkg.svc/Method`.
+//
+// message MethodRequest {
+//   string method_foo = 1;
+//   string method_bar = 2;
+//   repeated int32 method baz = 3;
+//   ...
+// }
+//
+// This is the filter config(expressed in JSON format).
+// {
+//   "proto_descriptor_bin":{...},
+//   "extractions_by_method": {
+//      "/pkg.svc/Method":{
+//         "request_field_extractions":{
+//            "name_foo":{
+//               "header":"grpc-field-extraction-output1-bin"
+//            },
+//            "name_bar":{
+//               "header":"grpc-field-extraction-output1-bin"
+//            }
+//            "name_baz":{
+//               "header":"grpc-field-extraction-output2-bin"
+//            }
+//         }
+//      }
+//   },
+//   ...
+// }
+//
+//
+// During runtime, the filter receives the following `MethodRequest` message(expressed in JSON format).
+// {
+//   name_foo: "val_foo",
+//   name_bar: "val_bar",
+//   name_baz: [123, 456]
+// }
+//
+// The filter will injects the following 2 request headers(expressed in JSON format).
+//
+// rpc-field-extraction-output1-bin: {
+//   "name_foo":[
+//      "val_foo"
+//   ],
+//   "name_bar":[
+//      "val_bar"
+//   ],
+//
+// rpc-field-extraction-output2-bin: {
+//   "name_baz":[
+//      "123"
+//   ]
+//   "name_baz":[
+//      "456"
+//   ]
+//
+
+message GrpcFieldExtractionConfig {
+  oneof descriptor_set {
+    option (validate.required) = true;
+
+    // Supplies the filename of
+    // :ref:`the proto descriptor set
+    string proto_descriptor = 1;
+
+    // Supplies the binary content of
+    // :ref:`the proto descriptor set
+    bytes proto_descriptor_bin = 2;
+  }
+
+  // Specify the extraction info.
+  // The key is the fully qualified gRPC method name.
+  // `/$package.$Service/$Method`, like
+  // `/endpoints.examples.bookstore.BookStore/GetShelf`
+  //
+  // The value is the field extractions for individual gRPC method.
+  map<string, FieldExtractions> extractions_by_method = 4;
+}
+
+// This message can be used to support per route config approach later even
+// though the Istio doesn't support that so far.
+message FieldExtractions {
+  // The field extractions for requests.
+  // The key is the field path within the grpc request.
+  // For example, we can define `foo.bar.name` if we want to extract
+  // Request.foo.bar.name.
+  // message Request {
+  //   // The namespace in which the Workspace should be created.
+  //   Foo foo = 1;
+  // }
+  //
+  // message Foo {
+  //   Bar bar = 1;
+  // }
+  // message Bar {
+  //   string name = 1;
+  // }
+  map<string, RequestFieldExtraction> request_field_extractions = 1;
```

Comment:
Would it make more sense to pre-process the extraction path on the control plane? Instead of sending the descriptor, use field numbers (and names for JSON) as the field path instead? it might be much faster with a SAX parser than trying to instantiate a dynamic message.

Reply:
Hi Kuat, IIUC, you want to emphasize not init the proto but directly parse from binary. In fact, we will do that(for details see the [library](https://github.com/grpc-ecosystem/proto-field-extraction/blob/main/src/field_extractor/field_extractor.cc) we are gonna rely on). Parsing proto descriptors + field name VS field names won't make difference? In fact, IMHO the former one gives better config readability.

Reply:
Great that it's an online parser. Proto descriptor might be OK, it's just rather large if all you want is one field path. From what I remember, it can easily go into MBs.

Reply:
IMHO, proto descriptor is a must if you want to read some fields from protobuf message binary. Only knowing path(field index) isn't enough.

ACK on `From what I remember, it can easily go into MBs.`.

Reply:
Ok, I think it's fine to load the descriptor since the intermediate types need to be known for parsing. For a top-level fields all you need to know is a field number.